This folder contains scripts to visualize and understand the input data (cfDNA fragment coordinates and Enformer predictions) better. It has scripts to do dimensionality reduction, train models (like logistic regression) and explore the separability (ability to separate donor- and recipient-derived fragments) of the feature sets used for training the Deep Learning models. 

### Description and purpose of all the scripts in this folder. 

| S.No | Script name | Description of contents | Purpose |
|------|-------------|-------------------------|---------|
| 1 | [coverage_plot_script.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/coverage_plot_script.ipynb) | Plot cfDNA fragment coverage (the number of fragments in a particular region) for chromosome 1 | To check whether donor- and recipient-derived fragments have coverage peaks in different regions of the genome (verify hypothesis behind the project) |
| 2 | [create_donor_recipient_sample_table.py](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/create_donor_recipient_sample_table.py) | Create a table of number of donor and recipient samples for each patient. | To verify the extent of class imbalance. |
| 3 | [enformer_output_initial_analysis.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/enformer_output_initial_analysis.ipynb) | Initial analysis like dimensionality reduction (using PCA and TSNE), and logistic regression with the 5,313 Enformer tracks for a sample as features. | To assess the potential of the 5,313 Enformer tracks as features for separating donor-derived fragments from recipient-derived fragments. |
| 4 | [enformer_output_initial_analysis_tsne_umap.py](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/enformer_output_initial_analysis_tsne_umap.py) | Python script to do TSNE and UMAP dimensionality reduction on Enformer epigenetic tracks. | Similar function as enformer_output_analysis.ipynb, but in py format for running with srun outside Jupyter notebook |
| 5 | [generate_enformer_tracks_chr1_reference_genome.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/generate_enformer_tracks_chr1_reference_genome.ipynb) | Jupyter test notebook for generating Enformer tracks all regions in chromosome 1 | - |
| 6 | [generate_enformer_tracks_chr1_reference_genome.py](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/generate_enformer_tracks_chr1_reference_genome.py) | Script which breaks the chromosome 1 into fragments of desired input length for Enformer, creates Enformer output and stores them in H5PY files. | To serve as an input for plotting in plot_tracks_reference_genome.ipynb. |
| 7 | [length_distribution_plot.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/length_distribution_plot.ipynb) | Makes length distribution plots for donor- and recipient-derived fragments. | To verify whether length as a feature has enough separability for use in classifying fragments as donor- and recipient-derived. |
| 8 | [patient_level_analysis.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/patient_level_analysis.ipynb) | Generates a plot of % dd-cfDNA calculated from true labels for various days post transplant for all patients. The plot also indicates whether at that time point, clinical symptoms of rejection were observed for the patient. | To verify whether the levels of % dd-cfDNA is a good indicator transplant rejection (measured as clinical signs of rejection). |
| 9 | [plot_enformer_track_ranges.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/plot_enformer_track_ranges.ipynb) | Plot all the Enformer track values, separated by types DNAse, CHIP-seq, ATAC-seq and CAGE for some samples. | To check the average ranges for different Enformer tracks and to understand whether normalization is required. |
| 10 | [plot_simulated_enformer_track_distribution.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/plot_simulated_enformer_track_distribution.ipynb) | Plot simulated data with varying percentages of samples and features containing augmented signals | To visualize the difference in the effect of adding signal to donor-derived and recipient-derived samples. |
| 11 | [plot_tracks_reference_genome.ipynb](https://github.com/vmadhupreetha/fragmentomics/blob/master/data_analysis/plot_tracks_reference_genome.ipynb) | Plot the DNAse track of one lung and blood cell type, for the entire chromosome 1 | To test the hypothesis on whether donor- and recipient-derived fragments originate from different regions of the genome. |

